docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,317 @@
+
+### Prerequisites
+
+We'll have to complete several independent prerequisites to realize this
+functionality, including a few long standing feature requests. These would be
+fully described in separate proposals.
+
+#### Shared PID Namespace
+
+Tracked in [#1615](https://issues.k8s.io/1615), a shared PID namespace became
+available in Docker 1.12 and requires plumbing through the docker drivers,
+extending the infra container to reap orphaned zombies, and [rolling
+out](https://pr.k8s.io/37404).
+
+#### Adding a Container to a Pod Spec
+
+Tracked in [#37838](https://issues.k8s.io/37838), Kubernetes disallows almost
+all modification to a pod spec as a conservative default and because arbitrary
+changes could have complex repercussions on resource accounting and quality of
+service. The most straightforward way to add a container to a running pod is to
+modify the pod spec.
+
+We will enable a container to be added to the list of containers in a pod spec
+so long as the new container does not specify resource requirements. Removing a
+container is not allowed.
+
+A prohibition on changing resources simplifies scheduling, but it means a pod's
+QoS class could drop from *Guaranteed* to *Burstable*. This is an acceptable
+consequence of modifying a running pod. It means a pod could potentially be
+evicted during debugging, particularly if there are insufficient resources to
+allow for the troubleshooting process.
+
+Admission controllers will watch for this operation in order to approve the
+images from the modified pod spec. Containers can be added regardless of pod
+status to allow for troubleshooting containers in states like
+`CrashLoopBackoff`.
+
+##### Alternative: Create a New Container Type
+
+Rather than modifying the default list of containers, we could create an
+additional list of containers in the spec similar to `InitContainers`. This
+would reduce the complexity of modifying the default containers list, but we
+should avoid a proliferation of container types unless there's a compelling
+reason. A mutable container list is a better long-term goal even if it is more
+difficult.
+
+#### Container Volumes
+
+Covered by [#831](https://issues.k8s.io/831), we'll want to be able to make
+volumes and eventually root filesystems of the other containers in the pod
+available to the troubleshooting container. This is a complex issue and covered
+in a separate proposal.
+
+When `kubectl` generates a patch to add a container to the pod spec, it will
+include configuration to do the equivalent of a "volumes from" the other
+containers in the pod, depending on how [#831](https://issues.k8s.io/831) is
+implemented.
+
+#### Container runOnce
+
+In order to support short-lived processes (e.g. troubleshooting sessions,
+automation) from the general pod spec container list we need to be able to
+override the pod's `restartPolicy` for the newly added container so it is not
+restarted when it exits. When a "run once" container exits, it will transition
```

Comment:
Yes, it does by design in order to meet the requirement that changes can be audited by existing tools. Alternatively we could garbage collect and surface the changes in some other field of Pod/PodStatus, but tools would have to be updated to check a new field.

I think @derekwaynecarr added this requirement, but I may have misunderstood it.